### PR TITLE
Enabled amin & amax for float16 & bfloat16

### PR DESCRIPTION
**Pull request by imaginary-person**
### SUMMARY
1. Enabled `amax` & `amin` for `float16` & `bfloat16` dtypes for both CPU & CUDA.
2. Added `OpInfo`s for `amax` & `amin`.
3. Enabled `test_min_with_inf` & `test_max_with_inf` for both `float16` & `bfloat16`, as they also use `torch.amin` & `torch.amax` respectively.
4. Enabled `test_amax` & `test_amin` for `float16` but not for `bfloat16`, as comparison is done with `numpy`, which doesn't support `bfloat16`.

### CODE REVIEW

**Comment by imaginary-person** [comment]
Hello @mruberry & @anjali411, this PR enables `amax` & `amin` for `float16` & `bfloat16`. I'm unable to add you as reviewers, so can you please add yourselves?

Thank you!

**Comment by mruberry** [comment]
> Hello @mruberry & @anjali411, this PR enables `amax` & `amin` for `float16` & `bfloat16`. I'm unable to add you as reviewers, so can you please add yourselves?
> 
> Thank you!

@anjali411 was generously covering for me since I was on PTO last week; I've added the appropriate reviewers. We'll take a look soon.

File: test/test_reductions.py [hunk 1]
```diff
@@ -499,17 +499,17 @@ def test_all_any_empty(self, device):
         self.assertTrue(x.all())
         self.assertFalse(x.any())
 
-    @dtypesIfCUDA(torch.half, torch.float, torch.double)
-    @dtypes(torch.float, torch.double)
+    @dtypesIfCUDA(torch.half, torch.bfloat16, torch.float, torch.double)
```
Comment (heitorschueroff) [inline test/test_reductions.py:502]: Is it necessary to test on all these types given OpInfo tests for supported dtypes already? Do we expect different kernels to run for these types?

**Review by heitorschueroff — APPROVED** [review]
LGTM

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1462,6 +1477,26 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
...
            sample_inputs_func=sample_inputs_addmm),
+    OpInfo('amax',
+           op=torch.amax,
+           dtypes=all_types_and(torch.float16, torch.bfloat16, torch.bool),
+           dtypesIfCPU=all_types_and(torch.float16, torch.bfloat16, torch.bool),
+           dtypesIfCUDA=all_types_and(torch.float16, torch.bfloat16, torch.bool),
+           test_inplace_grad=False,
+           sample_inputs_func=sample_inputs_amax_amin,
+           skips=(
+               SkipInfo('TestCommon', 'test_variant_consistency_jit',
```
Comment (heitorschueroff) [inline torch/testing/_internal/common_methods_invocations.py:1488]: Could you add a comment with the error you get from this test?

File: test/test_reductions.py (hunk 1 above)
Comment (imaginary-person) [reply test/test_reductions.py:502]: @heitorschueroff, these tests seem to be necessary because `amax` & `amin` `OpInfo` tests don't test for functionality.

As mruberry [explained earlier](https://github.com/pytorch/pytorch/pull/51244#issuecomment-770517153),

> OpInfos are really helpful metadata that let the test suite generate a variety of tests, but they don't tell the test suite whether the function is working as expected or not (this is not entirely true, but for amin and amax we can assume it's the case). So there are also tests for amin and amax functionality in test_reductions.py

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:1488]: Oops, I forgot to add it. Thanks for catching this!

**Comment by imaginary-person** [comment]
Hello @heitorschueroff, I'm also working on enabling `argmin` & `argmax` for `float16` & `bfloat16`, and wrote their respective `OpInfo`s, but there doesn't seem to be [autograd support](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml) for them right now. Whenever you can, please advise if autograd support needs to be added for them, or if I should skip autograd-related tests by using `supports_autograd=False` in their respective `OpInfo`s. Thank you!

**Comment by mruberry** [comment]
Yep, if the operations don't support autograd then using the "supports_autograd" metadata is the right fix.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:1500]: Hello @heitorschueroff, I had added the comments. Thank you!

**Comment by mruberry** [comment]
Hey @imaginary-person, I'm catching up from vacation but will take a look at this soon!

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -635,6 +635,21 @@ def sample_inputs_gather(op_info, device, dtype, requires_grad):
                         0, torch.tensor(0, dtype=torch.int64, device=device))),
             )
 
+def sample_inputs_amax_amin(op_info, device, dtype, requires_grad):
+    test_cases = (
+        ((S, S, S), ()),
+        ((S, S, S), (1,)),
+        ((S, S, S), ((1, 2,),)),
+        ((S, S, S), (1, True,)),
+        ((), (0,)),
```
Comment (heitorschueroff) [inline torch/testing/_internal/common_methods_invocations.py:644]: Can you also add the scalar/no_args version as follows:
``((), tuple()),``

**Review by heitorschueroff — APPROVED** [review]
This is looking pretty good. Could you just address the one comment on the OpInfo and ping me when it's done so I can land it. Thank you for your contribution!

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:644]: Oops, I had missed that one! Thanks again for catching it!

**Comment by imaginary-person** [comment]
Hello @heitorschueroff, thanks again for reviewing the PR! I've added the missing test input. Thanks for catching that mistake!

**Review by heitorschueroff — COMMENTED** [review]
Thanks for adding the missing test. Unfortunately, that seems to have confused mypy. Adding type annotations to test_cases will fix it, you can do that as follows: 
``test_cases: Tuple[Tuple[Tuple[int, ...], Tuple[any, ...]], ...]``
@mruberry is there a better way then defining these quite long types?

**Comment by imaginary-person** [comment]
@heitorschueroff, thank you for following up! I'm sorry, I should've waited for the CI checks to complete yesterday before pinging you. I had expected them to pass, so I hadn't followed up on their results, but I'll keep that in mind in the future. 

The failure was limited to a particular line number, so I think using `()` instead of `tuple()` should fix the issue. If the CI checks would fail, I'll follow through on your advice as soon as I'd be able to test on a machine.

Thanks again!

**Comment by mruberry** [comment]
> Thanks for adding the missing test. Unfortunately, that seems to have confused mypy. Adding type annotations to test_cases will fix it, you can do that as follows:
> `test_cases: Tuple[Tuple[Tuple[int, ...], Tuple[any, ...]], ...]`
> @mruberry is there a better way then defining these quite long types?

These mypy failure for the test fixtures are frustrating, although we'll soon be able to write sample inputs more simply as generator expressions that should avoid this issue. If the failure persists we can check with our current mypy expert for a better solution.

**Comment by imaginary-person** [comment]
I tested it locally & the failure has been resolved. The current CI failures are unrelated to the change. I'll keep monitoring the CI checks for any related issues. Thank you!

**Comment by mruberry** [comment]
The XLA failure on this PR is an unrelated networking issue. The ROCm failure is in base (and has been fixed). So CI looks good.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -1478,7 +1494,28 @@ def _make_tensor_helper(shape, low=None, high=None):
                SkipInfo('TestCommon', 'test_variant_consistency_eager',
                         dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16, torch.float16)),),
            sample_inputs_func=sample_inputs_addr),
-
+    OpInfo('amax',
+           op=torch.amax,
```
Comment (imaginary-person) [inline torch/testing/_internal/common_methods_invocations.py:1498]: This line is redundant as op is `torch.amax` by default, anyway. This line would be removed in another PR.

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -1478,7 +1494,28 @@ def _make_tensor_helper(shape, low=None, high=None):
...
+           sample_inputs_func=sample_inputs_amax_amin,
+           skips=(
+               # Reference: https://github.com/pytorch/pytorch/issues/48978
+               SkipInfo('TestCommon', 'test_variant_consistency_jit',
+                        dtypes=[torch.bfloat16]),)),
+    OpInfo('amin',
+           op=torch.amin,
+           dtypes=all_types_and(torch.float16, torch.bfloat16, torch.bool),
+           dtypesIfCPU=all_types_and(torch.float16, torch.bfloat16, torch.bool),
+           dtypesIfCUDA=all_types_and(torch.float16, torch.bfloat16, torch.bool),
```
Comment (imaginary-person) [inline torch/testing/_internal/common_methods_invocations.py:1512]: The above two lines are redundant, as `dtypesIfCPU` and `dtypesIfCUDA` should only be specified, if they're different from `dtypes`. They have since been removed in another PR.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -1478,7 +1494,28 @@ def _make_tensor_helper(shape, low=None, high=None):
                SkipInfo('TestCommon', 'test_variant_consistency_eager',
                         dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16, torch.float16)),),
            sample_inputs_func=sample_inputs_addr),
-
+    OpInfo('amax',
+           op=torch.amax,
+           dtypes=all_types_and(torch.float16, torch.bfloat16, torch.bool),
+           dtypesIfCPU=all_types_and(torch.float16, torch.bfloat16, torch.bool),
+           dtypesIfCUDA=all_types_and(torch.float16, torch.bfloat16, torch.bool),
+           test_inplace_grad=False,
```
Comment (imaginary-person) [inline torch/testing/_internal/common_methods_invocations.py:1502]: `test_inplace_grad` was recently replaced by `supports_inplace_autograd`, but after #54460 would land,
an `OpInfo` for an op should have a `supports_inplace_autograd=False` entry only if it has an inplace variant that does not support autograd.
